compiler/next/lib/resolution/scope-queries.cpp

```diff
@@ -0,0 +1,529 @@
+  return doLookupInScope(context, scope, name, inUseEtc, checkedScopes, result);
+}
+
+/* TODO
+static bool lookupExprInScope(Context* context,
+                              const Scope* scope,
+                              const Expression* expr,
+                              UseImportOther inUseEtc,
+                              BorrowedIdsWithName& result) {
+  if (auto ident = expr->toIdentifier()) {
+    return lookupInScope(context, scope, ident->name(), inUseEtc, result);
+  }
+  if (expr->isDot()) {
+    assert(false && "TODO");
+  }
+  assert(false && "Case not handled");
+  return false;
+}*/
+
+
+struct ImportsResolver {
+  Context* context = nullptr;
+  const Scope* scope = nullptr;
+  ResolvedImportScope* resolvedImports = nullptr;
+
+  ImportsResolver(Context* context,
+                  const Scope* scope,
+                  ResolvedImportScope* resolvedImports)
+    : context(context), scope(scope), resolvedImports(resolvedImports)
+  { }
+
+  std::vector<std::pair<UniqueString,UniqueString>>
+  convertOneName(UniqueString name) {
+    std::vector<std::pair<UniqueString,UniqueString>> ret;
+    ret.push_back(std::make_pair(name, name));
+    return ret;
+  }
+
+  std::vector<std::pair<UniqueString,UniqueString>>
+  convertLimitations(const UseClause* clause) {
+    std::vector<std::pair<UniqueString,UniqueString>> ret;
+    for (const Expression* e : clause->limitations()) {
+      if (auto ident = e->toIdentifier()) {
+        UniqueString name = ident->name();
+        ret.push_back(std::make_pair(name, name));
+      } else if (auto as = e->toAs()) {
+        UniqueString name;
+        UniqueString rename;
+        if (auto symId = as->symbol()->toIdentifier()) {
+          name = symId->name();
+        } else {
+          assert(false && "TODO");
+        }
+        rename = as->rename()->name();
+        ret.push_back(std::make_pair(name, rename));
+      }
+    }
+    return ret;
+  }
+
+  // make note of use/import
+  void visit(const Use* use) {
+    bool isPrivate = true;
+    if (use->visibility() == Decl::PUBLIC)
+      isPrivate = false;
+
+    for (auto clause : use->useClauses()) {
+      // First, add the entry for the symbol itself
+      const Expression* sym = clause->symbol();
+
+      if (!sym->isIdentifier()) {
+        assert(false && "TODO");
+        //as->symbol();
+        //as->rename()->name();
+      }
+      auto id = sym->toIdentifier();
+      auto name = id->name();
+
+      BorrowedIdsWithName r;
+      bool foundSym = lookupInScope(context, scope, name, UIO_USE, r);
+      if (foundSym == false) {
+        context->error(use, "undeclared identifier %s", id->name().c_str());
+      } else {
+        if (r.moreIds != nullptr) {
+          context->error(use, "ambiguity in resolving %s", id->name().c_str());
+        }
+        resolvedImports->imported.push_back(
+            ImportedSymbols(r.id, ImportedSymbols::SYMBOL_ONLY,
+                            isPrivate, convertOneName(name)));
```

Comment:
This felt kinda redundant that an `ImportedSymbol` always include the symbol itself (?).

Reply:
If you have `use M` then we actually bring in `M` as a symbol as well as the contents of `M`. This code is sortof normalizing the use statements into the ImportedSymbols and as such it treats the way that `use M` brought in `M` as a separate action from bringing in the contents of M.

Reply:
Right, but what I meant was if you have `use M` or `import M` or any version of those, we always bring in `M`. In other words we always bring in the symbol. So, the `push_back` just below should be sufficient in signaling that `M` is in scope. Or at least, that's what I was expecting.

In any case, I don't think things will work just by removing this, and something else needs to be done when we are consuming `resolvedImports`. And I don't feel too strongly that this code should change.